### PR TITLE
fix reentrant call error

### PR DESCRIPTION
**Pull request by simonjiao**
* **make sure that you have executed all the following process and no errors occur**
  - [x] make fmt
  - [x] make lint
  - [x] make test

* **The major changes of this PR**

* **Extra documentations**

### CODE REVIEW

File: tools/devnet/env.sh [hunk 1]
```diff
@@ -15,7 +15,7 @@ BIN_CFG="${BIN_CFG:=$BIN_CFG_DEFAULT}"
 BIN="target/$BIN_CFG"
 
 # logs
-ABCI_LOG_LEVEL="abciapp=info,baseapp=info,account=info,ethereum=info,evm=info,eth_rpc=info"
+ABCI_LOG_LEVEL="abciapp=info,baseapp=debug,account=debug,ethereum=debug,evm=debug,eth_rpc=debug"
```
Comment (ws4charlie) [inline tools/devnet/env.sh:18]: Can we still leave default logging levels untouched?

File: src/components/contracts/baseapp/src/app.rs [hunk 2]
```diff
@@ -77,10 +77,20 @@ impl abci::Application for crate::BaseApp {
         if let Ok(tx) = convert_unchecked_transaction::<SignedExtra>(raw_tx) {
             let check_fn = |mode: RunTxMode| {
                 let ctx = self.retrieve_context(mode).clone();
-                if let Err(e) = self.modules.process_tx::<SignedExtra>(ctx, tx) {
-                    info!(target: "baseapp", "Transaction check error: {}", e);
-                    resp.code = 1;
-                    resp.log = format!("Transaction check error: {}", e);
+                let result = self.modules.process_tx::<SignedExtra>(ctx, tx);
+                match result {
+                    Ok(ac) => {
```
Comment (ws4charlie) [inline src/components/contracts/baseapp/src/app.rs:82]: can we rename this `ac` as `ar` to be consistent with the one in deliver_tx?

File: src/components/contracts/baseapp/src/app.rs [hunk 3]
```diff
@@ -143,7 +153,7 @@ impl abci::Application for crate::BaseApp {
             let ret = self.modules.process_tx::<SignedExtra>(ctx, tx);
             match ret {
                 Ok(ar) => {
-                    debug!(target: "baseapp", "deliver tx succeed result: {:?}", ar);
+                    debug!(target: "baseapp", "deliver tx with result: {:?}", ar);
```
Comment (ws4charlie) [inline src/components/contracts/baseapp/src/app.rs:156]: Also check ar.code the same way as we did in check_tx and print out error message would be better?

File: src/components/contracts/modules/ethereum/src/impls.rs [hunk 4]
```diff
@@ -228,6 +232,24 @@ impl<C: Config> App<C> {
...
         }
 
+        let (code, message) = match reason {
+            ExitReason::Succeed(_) => (0, String::new()),
+            ExitReason::Error(_) => (2, "EVM error".to_string()),
+            ExitReason::Revert(_) => {
+                let mut message =
+                    "VM Exception while processing transaction: revert".to_string();
+                if data.len() > 68 {
+                    let message_len = data[36..68].iter().sum::<u8>();
```
Comment (ws4charlie) [inline src/components/contracts/modules/ethereum/src/impls.rs:242]: Any link for this output data parsing format?

File: src/components/contracts/modules/ethereum/src/impls.rs [hunk 5]
```diff
@@ -228,6 +232,24 @@ impl<C: Config> App<C> {
             ));
         }
 
+        let (code, message) = match reason {
+            ExitReason::Succeed(_) => (0, String::new()),
+            ExitReason::Error(_) => (2, "EVM error".to_string()),
```
Comment (ws4charlie) [inline src/components/contracts/modules/ethereum/src/impls.rs:237]: 1) Shouldn't be 0, 1, 2, 3 ?
2) For Error(_), Fatal(_) cases, they contains detailed sub reasons. If we don't have idea to pass everything in detail. Is it a good idea to move line<219> to the line immediately after the match and print the reason object ONLY when code != 0? like:

if code != 0 
    info!(target: "ethereum", "evm execute result: reason {:?} status {:?} used_gas {}", reason, status, used_gas);

File: src/components/contracts/primitives/core/src/transaction.rs [hunk 6]
```diff
@@ -155,14 +155,9 @@ where
...
             Ok(res) => {
-                if res.code == 0 {
-                    Extra::post_execute(ctx, pre, &res)?;
-                    ctx.state.write().commit_session();
-                    ctx.db.write().commit_session();
-                } else {
-                    ctx.state.write().discard_session();
-                    ctx.db.write().discard_session();
-                }
+                Extra::post_execute(ctx, pre, &res)?;
```
Comment (ws4charlie) [inline src/components/contracts/primitives/core/src/transaction.rs:158]: Does this mean we are not discarding session if code falls in [1, 2, 3]?

**Review by ws4charlie — COMMENTED** [review]
I left some comments.

File: src/components/contracts/primitives/core/src/transaction.rs (hunk 6 above)
Comment (simonjiao) [reply src/components/contracts/primitives/core/src/transaction.rs:158]: yes. If we discards uncommitted data, the client will never obtain the status of transaction it just submitted.

**Pull request closed without merging**